src/libs/kata-types/src/annotations/mod.rs

```diff
@@ -12,3 +23,754 @@ pub mod crio;
+    "io.katacontainers.config.runtime.disable_new_netns";
+/// A sandbox annotation to specify how attached VFIO devices should be treated.
+pub const KATA_ANNO_CONF_VFIO_MODE: &str = "io.katacontainers.config.runtime.vfio_mode";
+
+/// A helper structure to query configuration information by check annotations.
+#[derive(Debug, Default, Deserialize)]
+pub struct Annotation {
+    annotations: HashMap<String, String>,
+}
+
+impl From<HashMap<String, String>> for Annotation {
+    fn from(annotations: HashMap<String, String>) -> Self {
+        Annotation { annotations }
+    }
+}
+
+impl Annotation {
+    /// Create a new instance of [`Annotation`].
+    pub fn new(annotations: HashMap<String, String>) -> Annotation {
+        Annotation { annotations }
+    }
+
+    /// Deserialize an object from a json string.
+    pub fn deserialize<T>(path: &str) -> Result<T>
+    where
+        for<'a> T: Deserialize<'a>,
+    {
+        let f = BufReader::new(File::open(path)?);
+        Ok(serde_json::from_reader(f)?)
+    }
+
+    /// Get an immutable reference to the annotation hashmap.
+    pub fn get_annotation(&self) -> &HashMap<String, String> {
+        &self.annotations
+    }
+
+    /// Get a mutable reference to the annotation hashmap.
+    pub fn get_annotation_mut(&mut self) -> &mut HashMap<String, String> {
+        &mut self.annotations
+    }
+
+    /// Get the value of annotation with `key` as string.
+    pub fn get(&self, key: &str) -> Option<String> {
+        let v = self.annotations.get(key)?;
+        let value = v.trim();
+
+        if !value.is_empty() {
+            Some(String::from(value))
+        } else {
+            None
+        }
+    }
+
+    /// Get the value of annotation with `key` as bool.
+    pub fn get_bool(&self, key: &str) -> Option<bool> {
+        if let Some(value) = self.get(key) {
+            let value = value.trim();
+            if value.parse::<bool>().is_err() {
+                warn!(sl!(), "failed to parse bool value from {}", value);
+            } else {
+                return Some(value.parse::<bool>().unwrap());
+            }
+        }
+
+        None
+    }
+
+    /// Get the value of annotation with `key` as u32.
+    pub fn get_u32(&self, key: &str) -> Option<u32> {
+        let s = self.get(key)?;
+        match s.parse::<u32>() {
+            Ok(nums) => {
+                if nums > 0 {
+                    Some(nums)
+                } else {
+                    None
+                }
+            }
+
+            Err(e) => {
+                warn!(
+                    sl!(),
+                    "failed to parse u32 value from {}, error: {:?}", s, e
+                );
+                None
+            }
+        }
+    }
+
+    /// Get the value of annotation with `key` as i32.
+    pub fn get_i32(&self, key: &str) -> Option<i32> {
+        let s = self.get(key)?;
+        s.parse::<i32>()
+            .map_or(None, |x| if x < 0 { None } else { Some(x) })
+    }
```

Comment:
1. You don't need to check `nums > 0` when `parse::<u32>` was already successful.
2. I don't think the warning text adds much value, and we shouldn't fail silently.
3. Can't we just parse as with `bool`? (In theory, this could be shortened even further then with macros, but I fully understand if that's not your taste :))

Reply:
fixed in next version